RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+
+Similarly, CoreFX will annotate methods like `String.IsNullOrEmpty` as handling `null`. This means that the most common of null-checking methods can be respected by flow analysis and reduce the amount of nullability assertions.
+
+##### Possibly supported: Pattern matching with wildcard
+
+```fsharp
+let len (str: string | null) =
+    match str with
+    | null -> -1
+    | _ -> str.Length // OK - 'str' is string
+```
+
+A slight riff that is less common is treating `str` differently when in a wildcard pattern's scope. To reach that code path, `str` must indeed by non-null, so this could potentially be supported.
+
+#### Possibly supported: If check
+
+```fsharp
+let len (str: string | null) =
+    if str = null then
+        -1
+    else
+        str.Length // OK - 'str' is a string
+```
+
+Although less common, this is a valid way to check for `null` and could be enabled by the compiler. We know for certain that `str` is a `string` in the `else` branch. The inverse check is also true.
+
+#### Possibly supported: After null check within boolean expression
+
+```fsharp
+let len (str: string | null) =
+    if (str <> null && str.Length > 0) then // OK - `str` must be null
+        str.Length // OK here too
+    else
+        -1
+```
+
+Note that the reverse (`str.Length > 0 && str <> null`) would give a warning, because we attempt to dereference before the `null` check. This would only hold with AND checks. More generally, if the boolean expression to the left of the dereference involves a `x <> null` check, then the dereference is safe.
+
+#### Likely unsupported: boolean-based checks for null that lack handlesnull decoration
+
+Generally speaking, beyond highly-specialized cases, we cannot guarantee non-nullability that is checked via a boolean. Consider the following:
+
+```fsharp
+let myIsNull item = isNull item
+
+let len (str: string | null) =
+    if myIsNull str then
+        -1
+    else
+        str.Length // WARNING: could be null
+```
+
+Although this simple example could potentially work, it could quickly get out of hand. Also, other languages that support nullable reference types (C# 8.0, Scala, Kotlin, Swift, TypeScript) do not do this. Instead, non-nullability is asserted when the programmer knows something will not be `null`.
+
+#### Likely unsupported: nested functions accessing outer scopes
+
+```fsharp
+let len (str: string | null) =
+    let doWork() =
+        str.Length // WARNING: maybe too complicated?
+
+    match str with
+    | null -> -1
+    | _ -> doWork() // But after this line is written it's fine?
+```
+
+Although `doWork()` is called only in a scope where `str` would be non-null, this may too complex to implement properly.
+
+**Note:** C# supports the equivalent of this with local functions. TypeScript does not support this.
+
+#### Asserting non-nullability
```

Comment:
Please add a section on the interaction with optional parameters.  e.g. can you declare

    member x.M(?arg: (string | null))

and can you declare an optional arg where the optionality is represented by null (likewise Nullable, an existing issue today)